### ISSUE TITLE
[Feature Request] Pass custom scroll event when shallow routing.

### ISSUE THREAD

**Issue by rockchalkwushock** [zeit/next.js#2520]
- [x] I have searched the [issues](https://github.com/zeit/next.js/issues) of this repository and believe that this is not a duplicate.

## Expected Behavior

More like desired behavior; but it would be nice if a customScroll event could be passed by the user when executing shallow routing if the default scroll event provided by `next` [here](https://github.com/zeit/next.js/blob/v3-beta/lib/router/router.js#L269) is not the desired effect.

## Current Behavior

At the moment if trying to create your own custom scroll events or if using a third party library the scroll events will either be ignored or the scroll event being used by `next` takes precedence leading to two scroll events happening.

## Steps to Reproduce (for bugs)
Please see this [repo](https://github.com/rockchalkwushock/custom-scroll) for demo & a proposed solution (albeit probably not the greatest idea)

## Context

Really wanting shallow routing for a website I'm building with a scrolling layout. I'm using the `react-scroll` library to achiever the scrolling layout; but so far have found no concrete method for using the customScroll events being called by the library with the shallow routing feature in `next`.

## Your Environment

| Tech | Version |
|------|---------|
| next  | 3.0.1-beta.8 |
| chrome  | 59.0.3071.115  |
| node  | 8.1.3  |
| macOS  | 10.12.5 |

**Comment by rockchalkwushock** [comment]
I'm willing to spearhead this since it's something that I would _really like_ :smile: :tada:

**Comment by timneutkens** [comment]
@rockchalkwushock could you please submit the table you added instead of the text to the issue template. I love it 😄 ❤️ 

Regarding your proposal, cc @arunoda

**Comment by rockchalkwushock** [comment]
@timneutkens so you need the table in the issue as text, sorry didn't quite understand what you meant. I'm working on this now btw.

**Comment by rockchalkwushock** [comment]
nevermind I understand now....and English is my first only language :joy:

**Comment by rockchalkwushock** [comment]
I've got this working, at least a rough copy anyways. As soon as I can get my text editor to quit making format changes that prevent me from committing I'll push the code and open a PR.

**Comment by rockchalkwushock** [comment]
@timneutkens 

Still fighting formatting issues with the example I made 🙄 , but I do have a question about the how to integrate these changes into the `<Link />` API. I realize we would need to add the new props to the component, but where I'm a little lost is how to update the code [here](https://github.com/zeit/next.js/blob/v3-beta/lib/link.js#L72) to reflect this addition 😕 .

You can see the changes I have made to the code base in these [commits](https://github.com/rockchalkwushock/next.js/commits/feat/customScroll). If you have any concerns about potential setbacks or better methodology for handling this please let me know I'm all ears and by no means a wiz a this.

**Comment by timneutkens** [comment]
> nevermind I understand now....and English is my first only language 😂

@rockchalkwushock drawback of me typing stuff on my phone :joy:

Regarding your questions, @arunoda could you have a look on if we even want this in the core 😄👌

**Comment by rockchalkwushock** [comment]
@timneutkens 

haha no prob man I understand, I'll standby for direction from @arunoda

**Comment by arunoda** [comment]
I think we can have a setting to turn-off the scrolling we use here.
We can do something like this:

```js
import Router from 'next/router'
Router.disableScrolling()
```

**Comment by rockchalkwushock** [comment]
@arunoda 

I think that would be good. No rework to the code revolving around the shallow routing or props needing to be added to the `<Link />` component. Would we just want `disableScrolling` to look for scrolling events and remove them?

_router.js_
```js
// ...
disableScrolling (as) {
  const [ , hash ] = as.split('#')
  const el = document.getElementById(hash)
  // I've been reading around and it looks as if there is
  // no method for checking if an element has an event listener attached.
  // This is currently experimental, maybe a future fix should it become standard:
  // https://developer.mozilla.org/en-US/docs/Web/API/Document/scrollingElement
  if (el) { // bad
    el.removeEventListener('scroll')
  }
}
```

**Comment by arunoda** [comment]
Actually, I don't like ^^^ code.

I am looking at something like this:

```js
disableScrolling () {
  this.dontScroll = true
}

scrollToHash (as) {
  if (this.dontScroll) return
  
  const [ , hash ] = as.split('#')
  const el = document.getElementById(hash)
  if (el) {
    el.scrollIntoView()
  }
}
```

**Comment by rockchalkwushock** [comment]
@arunoda 

I see so if the user passes `Router.disableScrolling()` at the top of the file, this code will run. Would the user need to do the same for using `<Link href='#foo' shallow />`, or would we prefer a prop be defined for that use case? My assumption is `donScroll` is a prop, but I want to make sure before I proceed. I undo my commits and implement this. I saw there is a test file for _router.js_ is that the location I should write the tests for this, & the `<Link />` API update?
